src/pixie.c

```diff
@@ -18,20 +18,138 @@ void pixie_format(const unsigned char *in, unsigned len, char *outbuf) {
 	int dh_small = get_dh_small();
+	int ret_val;
 
 	if(p->do_pixie) {
 		char cmd[4096];
 		snprintf(cmd, sizeof cmd,
 		"pixiewps -e %s -s %s -z %s -a %s -n %s %s %s",
 		p->pke, p->ehash1, p->ehash2, p->authkey, p->enonce,
 		dh_small ? "-S" : "-r" , dh_small ? "" : p->pkr);
-		printf("executing %s\n", cmd);
-		exit(system(cmd));
+		if (p->do_pixie == 2) {
+			ret_val = pixie_wrapper(cmd);
+			if (ret_val != EXIT_SUCCESS) {
+				/* <sys/wait.h> WEXITSTATUS() or div 256 */
+				exit(ret_val);
+			}
+		}
+		else {
+			printf("executing %s\n", cmd);
+			exit(system(cmd));
+		}
 	}
+}
+
+void pixie_free() {
 	PIXIE_FREE(authkey);
 	PIXIE_FREE(pkr);
 	PIXIE_FREE(pke);
 	PIXIE_FREE(enonce);
 	PIXIE_FREE(ehash1);
 	PIXIE_FREE(ehash2);
+	PIXIE_FREE(wrapper);
+	PIXIE_FREE(pin);
+}
+
+/**
+ * Call a wrapper to run pixiewps.
+ * 
+ * The wrapper could be implemented in any language which needs to follow the following conventions:
+ * 1) arguments are passed precisely as for pixiewps
+ * 2) on failure, non-zero exit status is returned
+ * 3) on success, exactly one line is printed to stdout, containing the found pin and a trailing newline character,
+ *    in case of null pin, only a newline character '\n' would be returned and exit status 0 is returned.
+ * 
+ * @params char* cmd The command line to call pixiewps with the arguments
+ * 
+ * @return int
+ */
+int pixie_wrapper(char *cmd) {
+	char *sh = NULL;
+	int ret_val = EXIT_FAILURE;
+
+	sh = (char*) malloc(strlen(pixie.wrapper) + strlen(cmd));
+	if (sh) {
+		/* Call the wrapper with the arguments passed */
+		strcpy(sh, pixie.wrapper);
+		strcat(sh, " ");
+		strcat(sh, cmd+9);
+		printf("executing %s\n", sh);
+		ret_val = pixie_popen(sh);
+		free(sh);
+		sh = NULL;
+	}
+
+	return ret_val;
+}
+
+/**
+ * Executing pixiewps optimized with popen(), set the discovered pin to pixie.pin if Pixiewps attack is successful
+ * 
+ * 1) on failure, non-zero exit status is returned
+ * 2) on success, print the pin and exit status 0 is returned
+ * 
+ * @params char* cmd The command line to call a wrapper file
+ * 
+ * @return int
+ */
+int pixie_popen(char *cmd) {
+	FILE *fp;
+	char *str_pin = NULL;
+	char msg[256];
+	int ret_val = EXIT_FAILURE;
+
+	fp = popen(cmd, "r");
+	if (fp != NULL) {
+		while (fgets(msg, 256, fp) != NULL) {
+			printf("%s", msg);
+			/* Capture pin if the wrapper not following the specifications */
+			if(strstr(msg, "[+] WPS pin: ") != NULL) {
+				str_pin = (char*) malloc(10);
+				if (str_pin) {
+					str_pin[0] = '\0';
+					if (strstr(msg, "empty") == NULL) {
+						strncpy(str_pin, msg+15, 8);
+						str_pin[8] = '\0';
+					}
+				}
+			}
+		}
+		ret_val = pclose(fp);
+	}
+	else {
+		perror("popen");
+	}
+
+	if(ret_val == EXIT_SUCCESS) {
+		/* In case of the wrapper not following the specifications */
+		if (str_pin) {
+			strcpy(msg, str_pin);
+		}
+		/* pin found, remove '\r' or '\n' */
+		while(strlen(msg)>0 && (msg[strlen(msg)-1]=='\n' || msg[strlen(msg)-1]=='\r')){
```

Comment:
calling strlen(msg) over and over is not very effective, and it unnecessarily makes the code longer. why not cache the result in a size_t variable ?

Reply:
I did not want to declare a variable. Okay, I'll follow your advice. Thank!